cryostatio/cryostat
# fix(credentials): store JMX session credentials in ThreadLocal

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Fixes: https://github.com/cryostatio/cryostat-web/issues/656

## Description of the change:
This adds a [`ThreadLocal`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/ThreadLocal.html) to the `CredentialsManager` containing a `Map<String, Credentials>`. The `String` is just the `targetId` of the current request, if any, and the `Credentials` are likewise the corresponding session credentials `X-JMX-Authorization` extracted from the API request context, if any. The `CredentialsManager`, and for now (needs refactor cleanup) the `ConnectionDescriptor`, reference this `ThreadLocal` to check for any session-provided credentials. If they are present then they are used and preferred over backend stored credentials. Once the current request is completed these credentials are cleared out of the `ThreadLocal Map`. This map is only expected to contain a singleton value currently and should only ever have one key corresponding to the current request's `targetId` URL parameter, but for testing purposes and request safety I used the `Map` to ensure credentials are not accidentally spilling between calls.

This concept is pretty standard in the "thread-per-request" model used by many webservers, where a new thread is spawned for each incoming client request, and that thread handles the entire execution lifespan, finally terminating after the response has been written. Here we use Vert.x which is reactive and processes requests using a main verticle thread and a worker pool. Most of our API handlers, in particular ones that may handle passthrough credentials, are handled by a worker in the Vert.x threadpool. Which exact thread handles the request is not predictable, but and depending on the exact execution path of that request, the thread may end up dropping the request and another thread (the same or another) may pick up the request later, if the request call stack contains something like `vertx.executeBlocking()` or `targetConnectionManager.executeConnectedAsync()`, either of which push the next unit of executable code into a work queue to be picked up by a worker thread. If the request execution is handed off between threads in this way then the session credentials held in the `ThreadLocal` will not be accessible and will appear to be `null` to the thread that picks up the request. If, however, all of these asynchronous delegation calls happen late in the call stack, ie after the point where the original thread has saved the session credentials to `ThreadLocal` and then picked them back up again for the last time, then this does not matter.

## Motivation for the change:
This allows API requests using the old JMX passthrough authentication mechanism, the `X-JMX-Authorization` header, to work as expected again including for requests that trigger executions in the stack in other classes like the `JvmIdHelper` that do not have a direct call from the API handler that can pass through the `RoutingContext`/`RequestParameters` object directly so that the session credentials can be extracted at the call site. As explained above, if all of the requests are structured and executed in the right way, then the session credentials can be stored in a "global" `ThreadLocal` so that as the worker thread processes the request, the session credentials can be retrieved from any point in the call stack, without having to pass the `RoutingContext` object through as a method call argument the whole way through the stack. For reviewers reading this, this can be thought of as somewhat analogous to a React `<Context>` where pieces of state are available to descendant components anywhere below the provider in the component hierarchy. The net effect, if the above assumptions hold, is that the passthrough credentials mechanism works as expected again, without any significant refactoring of the internal interfaces to allow passing the `RoutingContext` through every layer of the call stack, and without resorting to temporary storage of session credentials in the encrypted keyring database.

## How to manually test:
1. *Run CRYOSTAT_IMAGE=quay.io... sh smoketest.sh...*
2. Open the `web-client` or use `curl`/`httpie` and exercise as many API calls as possible, without using stored credentials. In the `web-client` this would mean deleting any previously stored credentials, going to `/settings` -> `Advanced` and setting the credentials storage to `Session (Browser Memory)`. Use browser dev tools to inspect outgoing network requests for ex. active recordings queries and ensure that the `X-JMX-Authorization` credentials passthrough header is included in the request and that the request succeeds.
